### PR TITLE
feat(consensus/grandpa): Add warp sync proof verification

### PR DESCRIPTION
**Pull request by dimartiro**
## Changes

Implement warp sync proof verification.
The warp proof is verified by traversing the warp proof fragments (with the block headers and justifications), then we verify the justifications against the authorities based on the genesis authorities and the scheduled changes.
If we are able to verify all the fragments, then the warp proof is valid.

- [x] Add warp sync proof `verify` method
- [x] Minor fixes in proof generation
- [x] Use right grandpa types
- [x] Tests
    - [x] Improve tests
    - [x] Test both proof generation and verification

[Reference parity code](https://github.com/paritytech/polkadot-sdk/blob/4edb21982befcce000537f13e31a87a64480ab38/substrate/client/consensus/grandpa/src/warp_proof.rs#L287-L312)

Note: I'd rather to not use generics in this initial phase because that will imply more changes (eg: in `BlockState` and `GrandpaState`) we can revisit this later to make the structs generic.

## Tests

```sh
make test
```

## Issues

#4237

### CODE REVIEW

File: dot/network/warp_sync.go [hunk 1]
```diff
@@ -8,18 +8,34 @@ import (
...
 	libp2pnetwork "github.com/libp2p/go-libp2p/core/network"
 	"github.com/libp2p/go-libp2p/core/peer"
 )
 
 const MaxAllowedSameRequestPerPeer = 5
 
+type WarpSyncVerificationResult struct {
+	SetId         grandpa.SetID
+	AuthorityList grandpa.AuthorityList
+	Header        generic.Header[uint64, hash.H256, runtime.BlakeTwo256]
```
Comment (EclesioMeloJunior) [inline dot/network/warp_sync.go]: I think that the type for a block header number is uint (I know that for 64bit OS it is the same as uint64, but we want to standardise that for Gossamer)

File: dot/network/warp_sync.go [hunk 2]
```diff
@@ -8,18 +8,34 @@ import (
...
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime/generic"
 	"github.com/ChainSafe/gossamer/lib/common"
 	libp2pnetwork "github.com/libp2p/go-libp2p/core/network"
 	"github.com/libp2p/go-libp2p/core/peer"
 )
 
 const MaxAllowedSameRequestPerPeer = 5
 
+type WarpSyncVerificationResult struct {
```
Comment (EclesioMeloJunior) [inline dot/network/warp_sync.go:21]: would it be encoded/decoded as a P2P message?

Comment (dimartiro) [reply dot/network/warp_sync.go:21]: Nop, the p2p message is the `WarpSyncProof`, the `WarpSyncVerificationResult` is just the result of the verification function but we aren't going to send it in the network

File: internal/primitives/consensus/grandpa/grandpa.go [hunk 3]
```diff
@@ -55,6 +55,11 @@ type SignedMessage[H, N any] grandpa.SignedMessage[H, N, AuthoritySignature, Aut
 // Commit is a commit message for this chain's block type.
 type Commit[H, N any] grandpa.Commit[H, N, AuthoritySignature, AuthorityID]
 
+type ScheduledChange[N runtime.Number] struct {
```
Comment (timwu20) [inline internal/primitives/consensus/grandpa/grandpa.go:59]: Can you add a comment?

**Review by timwu20 — CHANGES_REQUESTED** [review]
Do you mind moving the `internal/client/consensus/grandpa/warp_sync.go` file to `lib/grandpa` package? I think this is more inline with the "old" way of doing this with the local interfaces for the older block `BlockState` and `GrandpaState`.  I'd like to keep things more analogous to  substrate in the `internal/primitives` and `internal/client` code.

File: internal/primitives/consensus/grandpa/grandpa.go (hunk 3 above)
Comment (dimartiro) [reply internal/primitives/consensus/grandpa/grandpa.go:59]: Done!

**Comment by dimartiro** [comment]
> Do you mind moving the `internal/client/consensus/grandpa/warp_sync.go` file to `lib/grandpa` package? I think this is more inline with the "old" way of doing this with the local interfaces for the older block `BlockState` and `GrandpaState`. I'd like to keep things more analogous to substrate in the `internal/primitives` and `internal/client` code.

Moved!

File: lib/grandpa/warp_sync.go [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+		return nil, fmt.Errorf("getting best block header: %w", err)
+	}
+
+	if beginBlockHeader.Number > lastFinalizedBlockHeader.Number {
+		return nil, errStartBlockNotFinalized
+	}
+
+	authoritySetChanges, err := p.grandpaState.GetAuthoritiesChangesFromBlock(beginBlockHeader.Number)
+	if err != nil {
+		return nil, err
```
Comment (EclesioMeloJunior) [inline lib/grandpa/warp_sync.go]: I would suggest adding context to the returned error so when debugging makes it easy to trace a possible problem.

```suggestion
		return nil, fmt.Errorf("getting auth changes: %w", err)
```

File: lib/grandpa/warp_sync.go [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+				if err != nil {
+					return nil, err
+				}
+
+				scheduledChange, err := consensusDigest.Value()
+				if err != nil {
+					return nil, err
+				}
+
+				parsedScheduledChange, _ := scheduledChange.(types.GrandpaScheduledChange)
```
Comment (EclesioMeloJunior) [inline lib/grandpa/warp_sync.go:308]: I know that block should not contains a scheduled change and a forced change, however, for sanity reasons can we check this casting using the `ok` value so we ensure the we only return a `types.GrandpaScheduledChange`?

**Review by EclesioMeloJunior — APPROVED** [review]
I made few comments, but overall looks good, nice work 🚀

File: lib/grandpa/warp_sync.go (hunk 5 above)
Comment (dimartiro) [reply lib/grandpa/warp_sync.go:308]: Sure! good suggestion

**Review by timwu20 — APPROVED** [review]

**Pull request merged**